selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,425 @@
+use super::*;
+use crate::ast_util::{
+    range,
+    scopes::{ScopeManager, Variable},
+    HasSideEffects,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    convert::Infallible,
+    fmt::{self, Display},
+    usize,
+};
+
+use full_moon::{
+    ast::{self, Ast},
+    visitors::Visitor,
+};
+use id_arena::Id;
+
+pub struct MismatchedArgCountLint;
+
+impl Rule for MismatchedArgCountLint {
+    type Config = ();
+    type Error = Infallible;
+
+    fn new(_: Self::Config) -> Result<Self, Self::Error> {
+        Ok(MismatchedArgCountLint)
+    }
+
+    fn pass(&self, ast: &Ast, _context: &Context) -> Vec<Diagnostic> {
+        let scope_manager = ScopeManager::new(ast);
+
+        // Firstly visit the AST so we can map the variables to their required parameter counts
+        let mut definitions = HashMap::new();
+        let mut definitions_visitor = MapFunctionDefinitionVisitor {
+            scope_manager: &scope_manager,
+            definitions: &mut definitions,
+            blacklisted_variables: HashSet::new(),
+        };
+        definitions_visitor.visit_ast(&ast);
+
+        let mut visitor = MismatchedArgCountVisitor {
+            mismatched_arg_counts: Vec::new(),
+            scope_manager,
+            definitions,
+        };
+
+        visitor.visit_ast(&ast);
+
+        visitor
+            .mismatched_arg_counts
+            .iter()
+            .map(|mismatched_arg| {
+                Diagnostic::new_complete(
+                    "mismatched_arg_count",
+                    format!(
+                        "this function takes {}",
+                        mismatched_arg
+                            .parameter_count
+                            .to_message(mismatched_arg.num_provided)
+                    ),
+                    Label::new_with_message(
+                        mismatched_arg.call_range,
+                        mismatched_arg.parameter_count.to_string(),
+                    ),
+                    Vec::new(),
+                    vec![Label::new_with_message(
+                        mismatched_arg.function_definition_range,
+                        "note: function defined here".to_owned(),
+                    )],
+                )
+            })
+            .collect()
+    }
+
+    fn severity(&self) -> Severity {
+        Severity::Error
+    }
+
+    fn rule_type(&self) -> RuleType {
+        RuleType::Correctness
+    }
+}
+
+struct MismatchedArgCount {
+    parameter_count: ParameterCount,
+    num_provided: PassedArgumentCount,
+    call_range: (usize, usize),
+    function_definition_range: (usize, usize),
+}
+
+#[derive(Clone, Copy, Debug)]
+enum ParameterCount {
```

Comment:
I sense this might be useful some other day, maybe we should move it into the ast util module?

Reply:
The `to_message` impl function is specific to this lint - should I remove it from the impl and make it a normal function here?

Reply:
Nevermind then.